### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by rogercloud**
Some refactor

### CODE REVIEW

File: curp/src/sync_manager.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            sync_chan,
+            comp_chan,
+            others,
+            log,
+        }
+    }
+
+    /// Run the `SyncManager`
+    pub(crate) async fn run(&mut self) {
+        let f = self.others.read().len().wrapping_div(2);
```
Comment (pzheng1025) [inline curp/src/sync_manager.rs:109]: give a long variable name.

File: curp/src/channel/key_mpsc.rs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        }
+    }
+
+    /// modify data in the message if necessary
+    pub(crate) fn keys(&self) -> &[K] {
+        self.inner.keys()
+    }
+
+    /// map the inner message to a closure
+    pub(crate) fn map_msg<R, F: FnOnce(&mut M) -> R>(&self, f: F) -> R {
```
Comment (pzheng1025) [inline curp/src/channel/key_mpsc.rs:63]: Use where to make signature clearer.

File: curp/src/channel/key_mpsc.rs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    /// Create a new sender
+    fn new(channel: Arc<Mutex<KeybasedChannel<MpscKeysMessage<K, M>>>>) -> Self {
+        Self {
+            inner: Arc::new(MpscKeybasedSenderInner { channel }),
+        }
+    }
+
+    /// Send a key and message to the channel.
+    /// If the channel closed, returns `None`; otherwise returns a `Notify`
+    #[allow(dead_code)]
```
Comment (pzheng1025) [inline curp/src/channel/key_mpsc.rs:168]: Is this still needed?

File: curp/src/channel/mod.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    /// Create a new `KeysMessageInner` structure
+    fn new(keys: Vec<K>, msg: M) -> Self {
+        Self {
+            keys,
+            msg: Arc::new(Mutex::new(msg)),
+        }
+    }
+
+    /// map the inner message to a closure
+    fn map_msg<R, F: FnOnce(&mut M) -> R>(&self, f: F) -> R {
```
Comment (pzheng1025) [inline curp/src/channel/mod.rs:42]: use where to make signature clearer.

File: curp/src/channel/key_mpsc.rs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+//! The channel has the following features:
+//! 1. The message coupled with keys, any two message with conflicting keys are conflicted.
+//! 2. Any message send to the channel is control by a ready token returned by the sender API.
+//! 3. Unready message is not received by the receiver and it blocks all the following conflict messages
+
+use std::cmp::Eq;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::hash::Hash;
+use std::sync::Arc;
+use std::time::Duration;
```
Comment (pzheng1025) [inline curp/src/channel/key_mpsc.rs:12]: keep the same convention for std lib use.

File: curp/src/channel/key_spmc.rs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    /// modify data in the message if necessary
+    pub(crate) fn keys(&self) -> &[K] {
+        self.inner.keys()
+    }
+
+    /// map the inner message to a closure
+    pub(crate) fn map_msg<R, F: FnOnce(&mut M) -> R>(&self, f: F) -> R {
```
Comment (pzheng1025) [inline curp/src/channel/key_spmc.rs:59]: use where.

File: curp/src/channel/key_mpsc.rs [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+{
+    /// Create a new sender
+    fn new(channel: Arc<Mutex<KeybasedChannel<MpscKeysMessage<K, M>>>>) -> Self {
+        Self {
+            inner: Arc::new(MpscKeybasedSenderInner { channel }),
+        }
+    }
+
+    /// Send a key and message to the channel.
+    /// If the channel closed, returns `None`; otherwise returns a `Notify`
```
Comment (pzheng1025) [inline curp/src/channel/key_mpsc.rs:167]: This comment is out of date.

File: curp/src/channel/key_mpsc.rs [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+        if self.proceeder.get(km).is_none() {
+            ready_pool.push_back(km.clone());
+            ready_cnt = 1;
+        }
+
+        while let Some(cur) = ready_pool.pop_front() {
+            self.inner.push_back(cur.clone());
+            if let Some(successor) = self.successor.remove(km) {
+                for s in successor {
+                    let (no_proceeder, has_pre) = if let Some(s_p) = self.proceeder.get_mut(&s) {
```
Comment (pzheng1025) [inline curp/src/channel/key_mpsc.rs:121]: Is it possible that `s` in successor but not in proceeder?

Comment (rogercloud) [reply curp/src/channel/key_mpsc.rs:121]: It's possible. If one node has successors, it's in successor structure. The same for preceeder.

File: curp/src/channel/key_mpsc.rs (hunk 3 above)
Comment (rogercloud) [reply curp/src/channel/key_mpsc.rs:168]: A good reminder, I need to scan the dead_code again.

File: curp/src/channel/key_mpsc.rs (hunk 8 above)
Comment (pzheng1025) [reply curp/src/channel/key_mpsc.rs:121]: if a node in another node's successor list, it must have at least one proceeder right?

File: curp/src/channel/key_spmc.rs [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        };
+
+        if ready_cnt > 0 {
+            self.new_msg_event.notify(ready_cnt);
+        }
+    }
+}
+
+/// The Sender for the `KeybasedChannel`
+pub(crate) struct SpmcKeybasedSender<K, M> {
```
Comment (pzheng1025) [inline curp/src/channel/key_spmc.rs:134]: !impl Sync

File: curp/src/channel/key_spmc.rs [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+}
+
+impl<K: Eq + Hash + Clone + ConflictCheck, M> SpmcKeybasedReceiver<K, M> {
+    /// Receive a message
+    /// Return (message, `msg_complete_sender`)
+    #[allow(dead_code)]
+    pub(crate) async fn recv(
+        &mut self,
```
Comment (pzheng1025) [inline curp/src/channel/key_spmc.rs:178]: change to &self

File: curp/src/channel/key_spmc.rs [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+            .recv()
+            .map(|msg| (msg, self.done_tx.clone()))
+    }
+
+    /// Receive a message with a `timeout`.
+    /// Return None if `timeout` hits,
+    /// otherwise return Some((message, `msg_complete_sender`))
+    #[allow(dead_code, clippy::expect_used, clippy::unwrap_in_result)]
+    pub(crate) async fn recv_timeout(
+        &mut self,
```
Comment (pzheng1025) [inline curp/src/channel/key_spmc.rs:198]: change &self

File: curp/src/channel/key_spmc.rs [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+//! The channel has the following features:
+//! 1. The message coupled with keys, any two message with conflicting keys are conflicted.
+//! 2. Any message send to the channel is control by a done token returned by the receiver API.
+//! 3. Undone message blocks all the following conflict messages
+
+use std::cmp::Eq;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::hash::Hash;
+use std::sync::Arc;
+use std::time::Duration;
```
Comment (pzheng1025) [inline curp/src/channel/key_spmc.rs:12]: keep the same convention for std lib use.

File: curp/src/sync_manager.rs [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+/// The message sent to the `SyncManager`
+pub(crate) struct SyncMessage<C>
+where
+    C: Command,
+{
+    /// Term number
+    term: TermNum,
+    /// Command, may be taken out
+    cmd: Option<Box<C>>,
```
Comment (pzheng1025) [inline curp/src/sync_manager.rs:47]: use Box to avoid copying big command?

File: curp/src/sync_manager.rs [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+                            SyncResponse::WrongTerm(_)
+                            | SyncResponse::EntryNotEmpty(_)
+                            | SyncResponse::PrevNotReady(_) => {
+                                // todo
+                            }
+                        }
+                    });
+
+                if synced_cnt == f {
+                    let _r = self
```
Comment (pzheng1025) [inline curp/src/sync_manager.rs:158]: Do we need to check Result?

File: curp/src/channel/key_mpsc.rs (hunk 8 above)
Comment (rogercloud) [reply curp/src/channel/key_mpsc.rs:121]: Yeah, you're right, they're a pair. I check all the possible cases to avoid using `unwrap`.

Comment (rogercloud) [reply curp/src/channel/key_mpsc.rs:121]: I'll replace it with a `unreachable` statement.

File: curp/src/channel/key_spmc.rs (hunk 9 above)
Comment (rogercloud) [reply curp/src/channel/key_spmc.rs:134]: Any type that not implements `Sync` is not syncable.

File: curp/src/sync_manager.rs (hunk 13 above)
Comment (rogercloud) [reply curp/src/sync_manager.rs:47]: Yes, to avoid data payload copy.

File: curp/src/sync_manager.rs (hunk 14 above)
Comment (rogercloud) [reply curp/src/sync_manager.rs:158]: Yes, this version channel returns error, we should abort sync manager and unrecoverable.

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**